### PR TITLE
[cluster-test] Integrate ClusterSwarmKube into cluster-test

### PR DESCRIPTION
**Pull request by ankushagarwal**
## Summary

* Use `ClusterSwarm#create_validator_and_fullnode_set` in setup() to setup the cluster
* Create a sample pod spec for launching cluster-test `cluster_test_pod_sample.yaml`
* Add `validator_instances` and `fullnode_instances` methods to `ClusterSwarm`
* Use `libra_init` image for init container
* Add flags `k8s_num_validators`, `k8s_fullnodes_per_validator` to main
* Update prometheus so that it works in k8s cluster
* Replace `thread::sleep` with `tokio::time::delay_for` in tx_emitter

## Test Plan

Ran cluster-test

### CODE REVIEW

File: testsuite/cluster-test/src/main.rs [hunk 1]
```diff
@@ -342,30 +352,69 @@ impl BasicSwarmUtil {
...
-            cluster.validator_instances().len(),
-            cluster.fullnode_instances().len(),
-            aws.workspace()
-        );
-        Self {
-            cluster,
-            aws,
-            prometheus,
-        }
+        Runtime::new().unwrap().block_on(async move {
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: We should probably just move `Rutnime` from ClustreTestRunner to ClusterUtil and use it here.

File: testsuite/cluster-test/src/main.rs [hunk 2]
```diff
@@ -342,30 +352,69 @@ impl BasicSwarmUtil {
...
+            let k8s = env::var("KUBERNETES_SERVICE_HOST").is_ok();
+            let aws = Aws::new(k8s);
+            let cluster_swarm = ClusterSwarmKube::new()
+                .await
+                .expect("Failed to initialize ClusterSwarmKube");
+            let cluster = if k8s {
+                cluster_swarm.delete_all().await.expect("delete_all failed");
+                info!("Successfully delete all old pods");
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: This could prob be debug, since there is `info!` statement right after

File: testsuite/cluster-test/src/main.rs [hunk 3]
```diff
@@ -342,30 +352,69 @@ impl BasicSwarmUtil {
...
+                    cluster_swarm.fullnode_instances().await,
+                )
+                .unwrap()
+            } else {
+                Cluster::discover(&aws).expect("Failed to discover cluster")
+            };
+            let cluster = if args.peers.is_empty() {
+                cluster
+            } else {
+                cluster.validator_sub_cluster(args.peers.clone())
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Why is this needed?

File: testsuite/cluster-test/src/main.rs [hunk 4]
```diff
@@ -607,6 +661,9 @@ impl ClusterTestRunner {
     pub fn cleanup_and_run(&mut self, experiment: Box<dyn Experiment>) -> Result<()> {
         self.cleanup();
         self.run_single_experiment(experiment, Some(self.global_emit_job_request.clone()))?;
+        Runtime::new()
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: You can use runtime from self

File: testsuite/cluster-test/src/tx_emitter.rs [hunk 5]
```diff
@@ -389,12 +387,16 @@ impl SubmissionWorker {
 async fn wait_for_accounts_sequence(
     client: &mut AdmissionControlClientAsync,
     accounts: &mut [AccountData],
+    instance: &Instance,
```
Comment (andll) [inline testsuite/cluster-test/src/tx_emitter.rs]: You can consider replacing `client` with`NamedAdmissionControlClient` here,  instead of adding extra param

File: testsuite/cluster-test/src/main.rs [hunk 6]
```diff
@@ -342,30 +352,69 @@ impl BasicSwarmUtil {
...
+            } else {
+                Cluster::discover(&aws).expect("Failed to discover cluster")
+            };
+            let cluster = if args.peers.is_empty() {
+                cluster
+            } else {
+                cluster.validator_sub_cluster(args.peers.clone())
+            };
+            let prometheus_ip = if k8s {
+                "libra-testnet-prometheus-server.default.svc.cluster.local"
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: We do share prometheus URL in logs so that ppl can look at it, but this URL does not seem to be shareable?

File: testsuite/cluster-test/src/main.rs (hunk 1 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Right now I want to leave this as is. We will need a huge refactor once we migrate completely to kubernetes, where I imagine we will have a single global Runtime initiated in main using `#[tokio::main]`

File: testsuite/cluster-test/src/main.rs (hunk 3 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: I'm not really sure, this was from before : https://github.com/libra/libra/blob/master/testsuite/cluster-test/src/main.rs#L347

File: testsuite/cluster-test/src/main.rs (hunk 6 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: This is only an internal URL which will be used by code. For external sharing, the url looks somthing like http://grafana.kush-k8s-testnet.aws.hlw3truzy4ls.com/ - I'll add this in a later diff because it requires quering the workspace from within k8s

File: testsuite/cluster-test/src/tx_emitter.rs (hunk 5 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/tx_emitter.rs]: Done

File: testsuite/cluster-test/src/main.rs (hunk 4 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Done

File: testsuite/cluster-test/src/main.rs (hunk 2 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Done

File: testsuite/cluster-test/src/cluster.rs [hunk 7]
```diff
@@ -48,6 +48,26 @@ impl Cluster {
...
 
+    fn get_mint_key_pair() -> KeyPair<Ed25519PrivateKey, Ed25519PublicKey> {
+        let seed = "1337133713371337133713371337133713371337133713371337133713371337";
+        let seed = hex::decode(seed).expect("Invalid hex in seed.");
+        let seed = seed[..32].try_into().expect("Invalid seed");
+        let mint_key = ValidatorConfig::new().seed(seed).build_faucet_client();
+        KeyPair::from(mint_key)
+    }
+
+    pub fn k8s(
```
[the remark this answers is not in the document]
Comment (ankushagarwal) [reply testsuite/cluster-test/src/cluster.rs]: Done

File: testsuite/cluster-test/src/main.rs (hunk 6 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Updated code so that we print correct public url in logs

File: testsuite/cluster-test/src/main.rs (hunk 1 above)
Comment (andll) [reply testsuite/cluster-test/src/main.rs]: sounds good re refactor later

I don't think we need to migrate to tokio::main, we want to configure runtime ourselves

File: testsuite/cluster-test/cluster_test_pod_sample.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+apiVersion: v1
```
Comment (andll) [inline testsuite/cluster-test/cluster_test_pod_sample.yaml]: How would we use this file? If you want to put it into repo, it probably should not contain smth like `kush_bbe30c38`

File: testsuite/cluster-test/src/aws.rs [hunk 9]
```diff
@@ -22,9 +22,13 @@ pub struct Aws {
 }
 
 impl Aws {
-    pub fn new() -> Self {
+    pub fn new(k8s: bool) -> Self {
         let ec2 = Ec2Client::new(Region::UsWest2);
-        let workspace = discover_workspace(&ec2);
+        let workspace = if k8s {
```
Comment (andll) [inline testsuite/cluster-test/src/aws.rs]: There is still workspace name even if it runs with k8s right? Why put 'k8s' string there?

File: testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs [hunk 10]
```diff
@@ -57,6 +60,11 @@ impl ClusterSwarmKube {
         image_tag: &str,
         delete_data: bool,
     ) -> Result<Vec<u8>> {
+        let cfg_fullnode_seed = if num_fullnodes > 0 {
```
Comment (andll) [inline testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: What will happen if you just put is unconditionally as `let cfg_fullnode_seed= CFG_FULLNODE_SEED`? Why it has to be "" if there is no full nodes?

File: testsuite/cluster-test/src/main.rs [hunk 11]
```diff
@@ -342,30 +354,79 @@ impl BasicSwarmUtil {
...
-        }
+        Runtime::new().unwrap().block_on(async move {
+            let k8s = env::var("KUBERNETES_SERVICE_HOST").is_ok();
+            let aws = Aws::new(k8s);
+            let cluster_swarm = ClusterSwarmKube::new()
+                .await
+                .expect("Failed to initialize ClusterSwarmKube");
+            let cluster = if k8s {
+                cluster_swarm.delete_all().await.expect("delete_all failed");
+                let image_tag = if let Some(image) = args.deploy.as_ref() {
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: It looks like only usage of `image_tag` below is referencing it, why do you need to clone it? you could keep it as reference
```
let image_tag = args.deploy. as_ref().unwrap_or("master)
```

File: testsuite/cluster-test/src/main.rs [hunk 12]
```diff
@@ -342,30 +354,79 @@ impl BasicSwarmUtil {
...
-        );
-        Self {
-            cluster,
-            aws,
-            prometheus,
-        }
+        Runtime::new().unwrap().block_on(async move {
+            let k8s = env::var("KUBERNETES_SERVICE_HOST").is_ok();
+            let aws = Aws::new(k8s);
+            let cluster_swarm = ClusterSwarmKube::new()
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Will that work on non-k8s workspace?

**Comment by ankushagarwal** [comment]
@andll Did you have any other comments on this PR?

File: testsuite/cluster-test/src/main.rs [hunk 13]
```diff
@@ -172,12 +178,14 @@ pub fn main() {
     let mut runner = ClusterTestRunner::setup(&args);
 
     if let Some(ref hash_or_tag) = args.deploy {
-        // Deploy deploy_hash before running whatever command
-        let hash = runner
-            .deployment_manager
-            .resolve(hash_or_tag)
-            .expect("Failed to resolve tag");
-        exit_on_error(runner.redeploy(&hash));
+        if !runner.k8s {
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Looks like with k8s we deploy every time, previously we were able to keep cluster as is and run additional commands on it.

Is it hard to preserve same thing for k8s setup? Discover cluster if no `--deploy` provided or deploy it if this command is present?

File: testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs [hunk 14]
```diff
@@ -294,4 +325,19 @@ impl ClusterSwarm for ClusterSwarmKube {
...
+        let data = Api::v1ConfigMap(self.client.clone())
+            .within(DEFAULT_NAMESPACE)
+            .get("workspace")
+            .await?
+            .data;
+        let workspace = data
+            .get("workspace")
+            .ok_or_else(|| format_err!("Failed to find workspace"))?;
+        Ok(format!(
+            "http://grafana.{}-k8s-testnet.aws.hlw3truzy4ls.com",
```
Comment (andll) [inline testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: Is there reason for having '-k8s-testnet' prefix here? Could we just keep same schema when domain is called "<workspace>. aws.hlw3truzy4ls.com"?

Comment (ankushagarwal) [reply testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: Without the suffix, there would be a conflict when same tf workspace name is used to create a cluster using the old testnet and the new k8s testnet.

File: testsuite/cluster-test/src/main.rs (hunk 13 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: We can do discovery similar to AWS, but I don't see any immediate use cases for it right now. We can do it later.

File: testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs (hunk 14 above)
Comment (andll) [reply testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: If someone created same workspace with k8s and tf, that would be their mistake right?
I don't feel like we should carry over some confusing prefix just for this strange corner case

If we can just make URL look exactly same as it is right now, that would probably just simplify things further

File: testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs (hunk 10 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: Because `docker-run-dynamic.sh` expects a fullnode network when `CFG_FULLNODE_SEED` is set. Setting `CFG_NUM_FULLNODES` to 0 gives an error.

File: testsuite/cluster-test/src/main.rs (hunk 11 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Done

File: testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs (hunk 14 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs]: Since both https://github.com/calibra/libra-ops/tree/master/terraform/aws-k8s-testnet and https://github.com/calibra/libra-ops/tree/master/terraform/testnet create AWS resources in the same project, and they have separate terraform workspaces, it is easy to create terraform workspaces with the same name in both these terraform projects. For eg: I would create the `kush` tf workspace name in both the projects for my personal clusters. Once we completely move away from the old testnet, it makes sense to remove the suffix, but for now I think it is ok.

File: testsuite/cluster-test/src/aws.rs (hunk 9 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/aws.rs]: We never use this Aws struct or its workspace with the kubernetes setup, so I have added a static string here.
 
With the kubernetes setup, the workspace is encapsulated by `ClusterSwarmKube`. workspace only affects the grafana baseurl and that is provided by the `get_grafana_baseurl` function

File: testsuite/cluster-test/src/main.rs (hunk 12 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: This will fail with the non-k8s environment. Fixing it.

Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Updated.

File: testsuite/cluster-test/cluster_test_pod_sample.yaml (hunk 8 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/cluster_test_pod_sample.yaml]: Removed file

File: testsuite/cluster-test/src/main.rs [hunk 15]
```diff
@@ -616,6 +674,7 @@ impl ClusterTestRunner {
         mut experiment: Box<dyn Experiment>,
         mut global_emit_job_request: Option<EmitJobRequest>,
     ) -> Result<()> {
+        self.wait_until_all_healthy()?;
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Why is this needed here? Should it be part of deploy code?

File: testsuite/cluster-test/src/main.rs [hunk 16]
```diff
@@ -887,18 +948,20 @@ impl ClusterTestRunner {
...
-                .await
-                .map_err(|e| {
-                    info!(
-                        "Failed to remove network effects for {}. Error: {}",
-                        instance, e
-                    );
-                })
-        });
-        self.runtime.block_on(join_all(futures));
+        if self.cluster_swarm.is_none() {
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: This should be run in k8s too right? Instances can persist between tests so we need to clean network effects?
(ok to add is as todo to an issue, if you prefer not to address it in this PR)

Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Once we support network effects in k8s, we can add cleanup.

File: testsuite/cluster-test/src/main.rs (hunk 15 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: This is because we do k8s deployment during `ClusterUtil::setup` where AWS discovery happens and `ClusterUtil` does not have the `wait_until_all_healthy` function.

Generally speaking, once we migrate completely to k8s, we will probably have to rethink the testing models and workflows which cluster test will support and at that point I presume we will have a major refactor of components. Till then I want to avoid any major refactor to support both environments as a lot of the refactored code will be thrown away.
